definitions/migrations/1.3.13/atomicmarket.sql

```diff
@@ -0,0 +1,353 @@
+        WHERE buyoffer.state = 3
+        GROUP BY buyoffer.market_contract, buyoffer.buyoffer_id
+;
+
+
+ALTER TABLE atomicmarket_stats_markets
+	ADD constraint atomicmarket_stats_markets_pkey
+		primary key (market_contract, listing_type, listing_id);
+
+
+CREATE INDEX atomicmarket_stats_markets_collection_name ON atomicmarket_stats_markets USING btree (collection_name);
+CREATE INDEX atomicmarket_stats_markets_buyer ON atomicmarket_stats_markets USING btree (buyer);
+CREATE INDEX atomicmarket_stats_markets_seller ON atomicmarket_stats_markets USING btree (seller);
+CREATE INDEX atomicmarket_stats_markets_price ON atomicmarket_stats_markets USING btree (price);
+CREATE INDEX atomicmarket_stats_markets_time ON atomicmarket_stats_markets USING btree ("time");
+CREATE INDEX atomicmarket_stats_markets_asset_id ON atomicmarket_stats_markets USING btree ("asset_id");
+CREATE INDEX atomicmarket_stats_markets_schema_name ON atomicmarket_stats_markets USING btree ("schema_name");
+CREATE INDEX atomicmarket_stats_markets_template_id ON atomicmarket_stats_markets USING btree ("template_id");
+
+
+
+DROP TABLE IF EXISTS atomicmarket_stats_markets_updates;
+CREATE TABLE atomicmarket_stats_markets_updates(
+    market_contract VARCHAR(12),
+    listing_type text,
+    listing_id BIGINT
+);
+
+
+DROP FUNCTION IF EXISTS update_atomicmarket_stats_markets_by_sale CASCADE;
+CREATE OR REPLACE FUNCTION update_atomicmarket_stats_markets_by_sale() RETURNS TRIGGER AS $$
+DECLARE
+    affects_stats_markets BOOLEAN;
+BEGIN
+    affects_stats_markets =
+        (TG_OP IN ('INSERT', 'UPDATE') AND NEW.final_price IS NOT NULL AND NEW.state = 3)
+        OR
+        (TG_OP IN ('DELETE', 'UPDATE') AND OLD.final_price IS NOT NULL AND OLD.state = 3);
+    IF (NOT affects_stats_markets)
+    THEN RETURN NULL;
+    END IF;
+
+    INSERT INTO atomicmarket_stats_markets_updates(market_contract, listing_type, listing_id)
+    VALUES (
+        CASE TG_OP WHEN 'DELETE' THEN OLD.market_contract ELSE NEW.market_contract END,
+        'sale',
+        CASE TG_OP WHEN 'DELETE' THEN OLD.sale_id ELSE NEW.sale_id END
+    );
+
+    RETURN NULL;
+END
+$$ LANGUAGE plpgsql;
+
+DROP TRIGGER IF EXISTS atomicmarket_sales_update_atomicmarket_stats_markets_tr ON atomicmarket_sales;
+CREATE TRIGGER atomicmarket_sales_update_atomicmarket_stats_markets_tr
+    AFTER UPDATE OR INSERT OR DELETE ON atomicmarket_sales
+    FOR EACH ROW
+    EXECUTE FUNCTION update_atomicmarket_stats_markets_by_sale();
+
+
+DROP FUNCTION IF EXISTS update_atomicmarket_stats_markets_by_auction CASCADE;
+CREATE OR REPLACE FUNCTION update_atomicmarket_stats_markets_by_auction() RETURNS TRIGGER AS $$
+DECLARE
+    affects_stats_markets BOOLEAN;
+BEGIN
+    affects_stats_markets =
+        (TG_OP IN ('INSERT', 'UPDATE') AND NEW.buyer IS NOT NULL AND NEW.state = 1 AND NEW.end_time < extract(epoch from now()))
```

Comment:
I'm also concerned here @fabian-emilius Mentioned an auction is completed when there were no new bids in the last 2 min. 

Should this be also considered here?

Reply:
The end time updates if there are new bids

Reply:
One thing which could be an issue that the filler does not necessarily need to be up to date. Would that work @andreme ? It can process old data and then the end time would have already passed but there are still new bids / price changes expected. I dont think we should use the timestamp in the filler code because of that. Can we provide a timestamp to the function which is the "current time" ?

Reply:
Probably unlikely, but the same could happen with the current MV, if it gets refreshed while filler is catching up.

I was wondering before why there is no state which indicates that the auction is closed?

Can I rely on the end_time changing when there's a new bid?

I'm not sure overall this is such a big issue, since this is just for the price stats? It might be out sync for a bit, that doesn't seem too bad.

There is another possible issue - I don't understand how the auctions work, but what happens if I make the last bid and the auction closes in 10 minutes - will the auction record be updated again (so that this trigger runs)?
If not, we might have to regularly throw all auctions that ended recently in the recalc queue...

Reply:
The blockchain does not fire anything when an auction ends so we cant have a dedicated state for it. An auction ended when state is pending and the filler timestamp is bigger than the end_time.

When there is a new bid, the price and the buyer updates. end_time only updates when there is a new bid less than 2 minutes before it ends because then the end_time will be extended.

Yes you are right that this behaviour can happen with a MV as well but it will be fixed after the next refresh.
I am not sure whether the row is removed and inserted (or updated) again if there is a new bid after the first refresh in this case. If thats the case, then its fine I guess because it should not happen too often. Is this the case @andreme ?

Reply:
I think I fund a good solution for this.
I've added a refresh_at to the update table. When an auction changes, 2 update records are created, one for right now and another one for 3 minutes after the auction ends. That way it gets quickly updated.
Good catch @alexsotocx 

https://github.com/pinknetworkx/eosio-contract-api/pull/64/commits/bd232213eed1c7c8e26f46c785ffec19c2e3acc4

Reply:
Mmh this 2 minutes is something that is configured and can be different. Also it continues to happen so if the auction was extended and there is another bid in the last second, it will be extended again. If you listen to updates to end_time, you can use that timestamp already because it will get updates on new bids.

But again, the current time is not related to what the filler is currently progressing. The filler can be always behind the blockchain state for 10 minutes which would make that 3 minutes a bit pointless because we are only comparing to current time but not the time the filler is currently processing.

Reply:
I dont think that `refresh_at` is a good solution because in the end it can be months old data which is currently processed where refresh_at obviously would be smaller than the current time but auction still may receive bids